### ISSUE TITLE
readline explodes on contact with invalid LC_CTYPE in environment

### ISSUE THREAD

**Issue by jclulow** [omniosorg/omnios-build#3537]
Users have occasionally reported core files from **bash** processes with a stack that looks like this:

```
 $ pstack core
core 'core' of 2651:    bash
 fffffc7feb7b8b7a kill () + a
 0000000000483fd6 ???????? ()
 00000000004841bd termsig_sighandler () + ad
 fffffc7feb7b2306 __sighndlr () + 6
 fffffc7feb7a4ee1 call_user_handler (b, 0, fffffc7fffdfcc40) + 1d1
 fffffc7feb7a51de sigacthandler (b, 0, fffffc7fffdfcc30) + ee
 --- called from signal handler with signal 11 (SIGSEGV) ---
 ffffffffffffffff ???????? ()
 fffffc7fef235cc3 _rl_init_eightbit () + 13
 fffffc7fef2141f6 rl_initialize () + b6
 0000000000495535 initialize_readline () + 6e5
 0000000000434dc5 ???????? ()
 0000000000437349 ???????? ()
 0000000000439677 ???????? ()
 000000000043d692 yyparse () + 442
 0000000000433e7a parse_command () + 3a
 0000000000433ff6 read_command () + 56
 0000000000434186 reader_loop () + b6
 000000000043344a main () + f0a
 0000000000431897 _start_crt () + 87
 00000000004317f8 _start () + 18
```

Anecdotally, this has been attributed to being an issue with locales; e.g., the user is using a locale for which we do not have support, or the user is using a _valid_ locale but on a machine without the locale packages installed.  Setting the **LANG** and **LC_** variables to empty strings or **C** makes it go away.

Setting aside the obviously unconscionable handling of **SIGSEGV**, I have dug in and, through many breakpoints and single steps, discovered the root cause here.  In `_rl_init_locale()`, the readline library obtains the value of **LC_CTYPE** from the environment and uses it uncritically in a call to **setlocale(3C)**:

```c
char *
_rl_init_locale (void)
{
  char *ret, *lspec;

  /* Set the LC_CTYPE locale category from environment variables. */
  lspec = _rl_get_locale_var ("LC_CTYPE");
  /* Since _rl_get_locale_var queries the right environment variables,
     we query the current locale settings with setlocale(), and, if
     that doesn't return anything, we set lspec to the empty string to
     force the subsequent call to setlocale() to define the `native'
     environment. */
  if (lspec == 0 || *lspec == 0)
    lspec = setlocale (LC_CTYPE, (char *)NULL);
  if (lspec == 0)
    lspec = "";
  ret = setlocale (LC_CTYPE, lspec);»   /* ok, since it does not change locale */

  _rl_utf8locale = (ret && *ret) ? utf8locale (ret) : 0;

  _rl_current_locale = savestring (ret);
  return ret;
}
```

(NB: I have trimmed the ifdefs out so that we're looking at the source that actually gets compiled.)

Note that we use **_rl_get_locale_var()** to look up the value of **LC_CTYPE** in the environment, which returns a string.  In the case of the most recent report, the value was (the invalid locale string) `UTF-8`.  Because it's found, we pass it straight to **setlocale()** a bit later, which fails to load the invalid locale and correctly returns **NULL**, having left the current locale setting alone.

A NULL check _is_ performed, after a fashion, on the next use of **ret**, when setting **_rl_utf8locale**!  But, then, we pass the NULL value to **savestring()** straight away after that.  It's a bespoke implementation of **strdup()**, essentially, which appears to get inlined when we build the library.  It uncritically passes the pointer _it_ gets to **strlen(3C)**, which is the thing that actually fails before the SIGSEGV handling destroys all the useful state.

I think we should probably patch readline to avoid doing the wrong thing here, and then perhaps try and take it upstream.  I looked into what happens on, say, an Ubuntu system, and it appears that two things are in play:

- their **/bin/bash** is not linked against libreadline dynamically, but instead uses the internal readline implementation they cart around in the bash source; this version of readline may well not have the defect
- the Ubuntu folks have created a magical mechanism for sanitising the environment, the [**locale-check** program](https://git.launchpad.net/ubuntu/+source/base-files/tree/locale-check.c), which is called from `/etc/profile.d/01-locale-fix.sh` as of [maas bug 1134036](https://bugs.launchpad.net/ubuntu/+source/maas/+bug/1134036)

The environment sanitisation is somewhat regrettable in that it is done totally silently.  No warning is emitted on stderr, so users are basically not informed that their system is probably misconfigured, or is at least not going to do what they want with respect to locales.

**Comment by jclulow** [comment]
Thanks to [https://m.unix.house/@ncommander@restless.systems](@ncommander@restless.systems) for bearing with my many questions about their environment.

See also this mailing list thread: https://illumos.topicbox.com/groups/omnios-discuss/T45307281d4837de9-Md4c3ab60dfea76fbad6febb0

**Comment by ptribble** [comment]
Adding another data point: On Tribblix, bash uses the internal readline implementation and doesn't crash. So it's just that, and not the magical locale-check, that's in play.

**Comment by jclulow** [comment]
> Adding another data point: On Tribblix, bash uses the internal readline implementation and doesn't crash. So it's just that, and not the magical locale-check, that's in play.

Right, I mention the **locale-check** business chiefly because I was surprised to find it and how it works -- and because it's probably masking similar mishandling of invalid locale settings in a lot of other software on Ubuntu systems.

**Comment by citrus-it** [comment]
It looks like this has been fixed upstream in https://ftp.gnu.org/gnu/readline/readline-8.2-patches/readline82-001

I've imported all of the available patches for readline 8.2 and opened PRs for the affected OmniOS versions. The fix will be in next week's releases, thanks for the report and putting the analysis work in!
